etc/opensc.conf.example.in

```diff
@@ -211,6 +211,44 @@ app default {
 		#can = 123456;
 	}
 
+	card_driver PIV-II {
+		# "piv_max_object_size"
+		# defines the buffer size used to read PIV objects
+		# Although NIST sp800-73-4 lists object sizes, these are not hard limits.
+		# Other PIV-like cards may have larger objects.
+		# Maximum: 65535
+		# Minimum: 16384
+		# Default: 16384
+		# piv_max_object_size = 16384;
+		# May be set via environment: PIV_MAX_OBJECT_SIZE=16384
```

Comment:
PIV specs define a very limited set of APDU, and objects are read using GET DATA.

I had looked at reallocating with GET DATA. You idea does not work without major changes to how get-response works. I did not want to touch that. There is no way the driver gets control to be able to realloc the buffer.  And the the size of the object is only contained in the the first few bytes of the object, and no way to know if object exist other then to use GET DATA.  The current code which was messy enough would do a GET DATA for first 8 byte, which causes 256 bytes to be read, then allocate a buffer and issue GET DATA again.
 
The way to avoid problems with code it the field was to make a parameter that could be changed before the next release which could set a new default.

Reply:
The problem is that PIV has dozens of objects, which are sometimes read only a single time. Your allocations are causing thousands of wasted KB of memory...

My suggestion for a different approach, are just that, suggestions. I'm happy to review more options.

Reply:
Disagree.

All the certificates have to be read early to determine if the certificate exist. The SPKI from the certificate is parsed to get key type, size and parameters. No other way know the key size or type other then to guess at it. 

These cards are not pkcs15, no directory of certs, keys or data objests.  

To emulate pkcs15 card-piv.c and pkcs15-piv.c  present to the user which certificates are present (it has already read the certificates object, extracted and (if needed) unzipped the certificate and cached both the certificate object and the certificate.

When a calling application asks for what objects are present, it presents to the caller that the other objects like fingerprint or iris image objects (that can be large and not read above) are **possibly** available. This then avoids reading thes large objects until the calling application request them.  For example pkcs15-tool can read an individual object using the label. 

If a calling PKCS11 application is "lazy" and does a C_FindObjects without specifying the type (cert, key, data), all objects will be read from card or from the cache. Firefox years ago was being "lazy" when it was only trying to find a specific certificate it had already requested but di not request just certificates hey changed that to at least only request certificates. This was also being done between the time a context specific login verify was sent to card and when the  sign operation was being done. This caused extra APDU to be sent to the card which the PIV would not do the sign operation. The circumvention at that time was to turn on PIN caching but which would not all the use of PIN PAD reader.  (Firefox eventual changes the code to find only certificates.)

I bring up these points, to show how complicated things can get. I have spent a lot of time trying to optimize I/O from card by caching at the cost of a few KB and make sure PIN PAD readers would work with with the sign key. 

If the object size is unknown a temp read buffer is allocated:
https://github.com/dengert/OpenSC/blob/PIV-4-extensions/src/libopensc/card-piv.c#L2888-L2890

Then later a correct size buffer will be allocated for the object, and the temp read buffer freed.
https://github.com/dengert/OpenSC/blob/PIV-4-extensions/src/libopensc/card-piv.c#L2950-L2960

So there is only one large read buffer and only while doing the GET DATA. The data being cached is the size of the object so no memory is wasted.

Reply:
> All the certificates have to be read early to determine if the certificate exist

By holding data in the object cache in card-piv.c you're essentially duplicating data that is already present in the PKCS#15 emulator. Also, you're using this for all kind of objects, not only frequently used stuff. For example, the discovery object has only a couple of bytes and you're extracting a single bit of information from that and then you put this information into yet an other memory location (`pin_preference`); nevertheless you're caching the discovery object forever in an 16KB chunk of memory, where it is never read again.

> These cards are not pkcs15, no directory of certs, keys or data objests.

Sure, but that's no reason to hold every single card object in memory. Even if you need some of that frequently without wanting to read the card, you should use OpenSC's file caching. If that file caching isn't flexible enough (you want to use volatile memory instead of files, you want to use it at the card driver leveln instead of the pkcs15 level, you want support for auto-zipping), please adjust OpenSC's file caching mechanism so that everyone can benefit from your change.

> I bring up these points, to show how complicated things can get.

I appreciate all you've done to make things work. Thank you! Neither you nor I will work on this project forever, so we need to make sure that future developers can figure out all the details. Hence, we need to design the APIs so that they are clear and reusable. If there is any pain regarding the data caching, please try to fix this in the related module.

> The data being cached is the size of the object so no memory is wasted.

For my Yubikey, my profile says that *all* allocations in card-piv.c account for around 2KB of memory (`pkcs11-tool --login --test`). With that single change in question, an additional 16KB are used. I didn't look at the details - from my understanding that should have been even way worse. I bring this up, because it's both easier to do better and should be solved more universally.

Reply:
> I appreciate all you've done to make things work. Thank you! Neither you nor I will work on this project forever, so we need to make sure that future developers can figure out all the details. 

Your welcome. And thanks for leading this project. 

As you said: "Neither you nor I will work on this project forever...". 

I first started using OpenSC in 2003, and joined the "OpenSC-devel@opensc.org" 03/22/2005 and wrote the 
pkcs15-gemsafeGPK.c for cards we were testing. And wrote card-piv.c and pkcs15-piv.c using beta testcards from several sources in 2005.  Started looking at PIV and Windows in 2004 while NIST 800-73 was still in draft.   Kerberos was at the heart Windows Active Directory authentication and PKINIT is still used with AD smart card login to AD.) 

I retired in 2015. I enjoyed programming and smartcards so have kept it up to this day.   It is not as enjoyable as it used to be especially trying to keep #2053 as a PR. It has a lot more then just SM. If you don't like the SM, comment out line 64 `#define ENABLE_PIV_SM`

The first computer I ever used was an IBM 1401 in 1967. It had 16,000 bytes of memory. Saving memory  was a big deal back then, but I am not going to spend any time on trying to save 16KB in the PIV driver.

So Frank, if you want to make changes to PIV go ahead, do what you want. I will try and follow along on what you do as time permits.

Reply:
16KB... that number made me curious, so I tested with both, `pkcs11-tool --test --login` and firefox (multiple authentications) with identical results. It turns out that the CHUI and Discovery are the only objects that get cached (grep for `added #`). Discovery is cached twice (?). None of the cached files ever produce a cache hit (grep for `found #`). Discovery is what you don't want to cache, because you're using it to check whether the applet is still active. CHUI is read exactly once, when the pkcs#15 object needs to be initialized with a serial number. This raises an even bigger question... Is the internal cache needed at all? Do you have different results?

Reply:
I noticed that the pkcs#15 emulator additionally requests objects 6 and 8 (sign and auth cert), which aren't on the card. In this case, sc_pkcs15_cache_file could also be used....

Reply:
I would like to mark this as a difference of option.  I am not concerned about using a few extra K of memory at the expense of doing more I/O  operations to the card. The 16K buffer cuts out the having to read the first part of an object to find its length then using GET DATA to read the full object. 

There is still this possibility:
> I had looked at reallocating with GET DATA. You idea does not work without major changes to how get-response works. I did not want to touch that. There is no way the driver gets control to be able to realloc the buffer. And the the size of the object is only contained in the the first few bytes of the object, and no way to know if object exist other then to use GET DATA. The current code which was messy enough would do a GET DATA for first 8 byte, which causes 256 bytes to be read, then allocate a buffer and issue GET DATA again.

Reply:
As a general idea, I'd like to avoid complexity in favor of maintainability, that's why...
- I wonder why we need the PIV internal caching mechanism if it doesn't produce any cache hits apart from the certificates that we already have a caching mechanism for.
- And I wonder why the *user* has to choose a technical parameter that she likely doesn't understand.

I don't mind wasting some KB, but if the implementation can be built a) without restrictions to the object size which at the same time b) removes distractions from the user, I think we should better build it that way.

Reply:
All the buffer size setup could be avoided if OpenSC in `apdu.c` supported GET_DATA (odd) where the the response contains the size of the data to be read. `apdu.c` could then dynamically (re)allocate then handle additional get responses. 

I never tried to submit such a PR as it is a such a low level that I never thought it would be accepted.

I am on vacation and don't have access to ISO docs. There are a number of APDUs like GET_DATA that support even and odd versions where odd version expects TLV  to give length. 

When piv driver was written in 2004 I don't believe we where caching certificates in pkcs15 level. The only way to tell if a PIV card has a key and its properties is read certificate and find SPKI to get size and type. 

The PIV driver internally reads DISCOVERY object  to test if PIV applet is currently selected  and compares response to cached DISCOVERY object if available. This was done as some early Yubico would lose login state if SELECT AID was used.

If you feel that you can can improve the caching or save some space, go ahead. I would appreciate if you base changes on #2053.

Reply:
GET DATA (0xCB) may be executed with offset, i.e. offset DO'54'. 0xCA can be combined with SELECT DATA (0xA5) to pass the offset. At least, that's what ISO 7816-4 says, I'm not sure what the different PIV implementations are supporting.

I already have #2448 waiting for this PR and I may add the removal of internal caching for simplification, but I agree that we should move forward.

We could settle this discussion, if you removed the config option `piv_max_object_size` and its environment variable and just *choose* `16384` as default value for the object size. For the `#define` of the object size, please add a TODO to make the read out of the data more flexible.

Reply:
The real solution to this is "apdu.c could then dynamically (re)allocate then handle additional get responses" The use of a  `piv_max_object_size` whould then not be needed.

Reply:
When reading an "PIV object" using GET_DATA (CB) the size of the unknown. PIV is not PKCS15.  The only way to determine the size of the object is to read it using GET_DATA. 

ISO 7816-4  says: "If bit b1 is set to 1 (odd INS code), payloads (if any) shall be encoded in BER-TLV (see 8.1)."  

NIST is following this.  800-73-4 says in Part 2, 3.1.2 GET DATA Card Command:
"Tag list '5C' M BER-TLV tag of the data object to be retrieved. See Table 3, Part 1." and 
"Data Field BER-TLV with the tag '53' containing in the value field of the requested data object".
 
The above also applies to SM.  In the ISO SM code in OpenSC, the payload is split up into small chunks and each is written or read as a separate SM APDU. But in the NIST case the payload is encoded as one chunk and sent to the card using COMMAND_CHAINING, which has as many chained APDUs as needed. Likewise the card can respond to an APDU (either SM or not) by sending back "6100". chunks are collected into a resp_buffer pre allocated by the application.  If this is with a SM response, the MAC covering the whole response is verified and then decrypted.
and returned in a buffer the size of the response. 

With command chaining and get response, the size of any payload is not limited to SC_MAX_APDU_RESP_SIZE (with or without SM) or limited by a reader or protocol. It is really only limited by card memory.  

Unfortunately, the interworking of OpenSC GET_RESPONSE does not allow for a way to look at the first BER-TLV to get total  size of the data to get an appropriately sized buffer.

What I would like to do is modify `apdu.c`  (when some `apdu->flag` is set)  to use the BER-TLV from the first date received to allocate/reallocate the response buffer. (There should also be some upper limit to avoid so DOS attack.) 
This would be used to avoid the use of `piv_max_object_size`.  

>Please either use a heuristic as I did (could also use some ASN.1 length to determine the length of an object) or just go with the maximum value. In any case, the user is likely to be unqualified to decide on the correct value.

The `piv_max_object_size` is more or less an upper limit, but it is also the size of the buffer used for most PIV GET_DATA operations.  It is also smaller then SC_MAX_EXT_APDU_RESP_SIZE used in the code example above which is 65k